### PR TITLE
chore(typing) Improve types for env()

### PR DESCRIPTION
**Pull request by markstory**
I'm working on config file changes and noticed that many of our calls to env() are flagged as invalid by pyright

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -25,7 +25,11 @@ def gettext_noop(s):
 socket.setdefaulttimeout(5)
 
 
-def env(key, default="", type=None):
+def env(
+    key: str,
+    default: Union[str, int, float, bool, None] = "",
+    type: Optional[Callable[[Any], Any]] = None,
+):
```
Comment (wmak) [inline src/sentry/conf/server.py:32]: Should we type the response too?

**Review by wmak — APPROVED** [review]
lgtm

File: src/sentry/conf/server.py (hunk 1 above)
Comment (evanh) [reply src/sentry/conf/server.py:32]: +1

Comment (markstory) [reply src/sentry/conf/server.py:32]: The return type would have to be Any :cry: We might be able to get better types with overloads though.

**Pull request merged**